test/end2end_test.go

```diff
@@ -6880,6 +6880,170 @@ func (s *httpServer) start(t *testing.T, lis net.Listener) {
+	return &transportRestartCheckCreds{}
+}
+func (c *transportRestartCheckCreds) OverrideServerName(s string) error {
+	return nil
+}
+
+type streamWithPeer struct {
+	stream testpb.TestService_FullDuplexCallClient
+	peer   *peer.Peer
+	id     string
+	conn   net.Conn
+}
+
+func (s *streamWithPeer) sendAndReceive() error {
+	if err := s.stream.Send(&testpb.StreamingOutputCallRequest{}); err != nil {
+		return fmt.Errorf("sending on %s: %v", s.id, err)
+	}
+	if _, err := s.stream.Recv(); err != nil {
+		return fmt.Errorf("receiving on %s: %v", s.id, err)
+	}
+	return nil
+}
+
+// Tests that the client transport drains and restarts when next stream ID exceeds
+// MaxStreamID. This test also verifies that subsequent RPCs use a new client
+// transport and the old transport is closed.
+func (s) TestClientTransportRestartsAfterStreamIDExhausted(t *testing.T) {
+	// Set the transport's MaxStreamID to 5 to cause connection to drain after 2 RPCs.
+	originalMaxStreamID := transport.MaxStreamID
+	transport.MaxStreamID = 5
+	defer func() {
+		transport.MaxStreamID = originalMaxStreamID
+	}()
+
+	ss := &stubserver.StubServer{
+		FullDuplexCallF: func(stream testpb.TestService_FullDuplexCallServer) error {
+			for {
+				if _, err := stream.Recv(); err != nil {
+					return err
+				}
+				if err := stream.Send(&testpb.StreamingOutputCallResponse{}); err != nil {
+					return err
+				}
+			}
+		},
+	}
+
+	creds := &transportRestartCheckCreds{}
+	if err := ss.Start(nil, grpc.WithTransportCredentials(creds)); err != nil {
+		t.Fatalf("starting stubServer: %v", err)
+	}
+	defer ss.Stop()
+
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	var streams []streamWithPeer
+
+	// Setting up 3 streams and calling Send() and Recv() once on each.
+	for i := 0; i < 3; i++ {
+		var p peer.Peer
+		s, err := ss.Client.FullDuplexCall(ctx, grpc.Peer(&p))
+		if err != nil {
+			t.Fatalf("creating FullDuplex stream: %v", err)
+		}
+		streamID := fmt.Sprintf("stream %v", i)
+		sp := streamWithPeer{stream: s, peer: &p, id: streamID}
+		streams = append(streams, sp)
+
+		if err := sp.sendAndReceive(); err != nil {
+			t.Fatal(err)
+		}
+	}
+	// Verifying only 2 connections are created so far.
+	if len(creds.connections) != 2 {
+		t.Fatalf("number of connections created: %v, want: 2", len(creds.connections))
+	}
+
+	// Verifying that streams on the first conn still works.
+	if err := streams[0].sendAndReceive(); err != nil {
+		t.Fatal(err)
+	}
+	if err := streams[1].sendAndReceive(); err != nil {
+		t.Fatal(err)
+	}
+
+	// The peer passed via the call option is set up only after the RPC is complete,
+	// which is why we wait here until Recv() returns EOF.
+	for i := 0; i < 3; i++ {
+		if err := streams[i].stream.CloseSend(); err != nil {
+			t.Fatalf("CloseSend() on %s: %v", streams[i].id, err)
+		}
+		for {
+			_, err := streams[i].stream.Recv()
+			if strings.Contains(err.Error(), "EOF") {
```

Comment:
Does `err == io.EOF` not work here?

Reply:
No it doesn't. The error returned is like `rpc error: code = Unknown desc = EOF` and does not match `io.EOF`

Reply:
According to https://pkg.go.dev/google.golang.org/grpc#ClientStream, `RecvMsg()` should return `io.EOF` when the stream completes successfully. Do you know why the return code here is wrapped as a status error with code `Unknown`?

Reply:
The FullDuplexCallF was set to return the err returned from `stream.Recv()`. Ive fixed it. It works now